### ISSUE TITLE
Specifying custom image for hasura on M1 mac devices

### ISSUE THREAD

**Issue by mwarger** [nhost/cli#76]
The current version of the hasura engine does not work on M1 Macs.  However, this image does: fedormelexin/graphql-engine-arm64:1.3.3.  I would like to be able to specify my own engine for running hasura in the docker-compose file, but it's deleted whenever the nhost dev command is stopped.  It would be nice if there was a config variable where we could override the docker image tag if needed.

**Comment by elitan** [comment]
Maybe we can use [`os.arch()`](https://nodejs.org/api/os.html#os_os_arch) to determine the CPU type and adjust the image based on that?

```
╭─eli@Johans-MacBook-Pro.local ~/tmp
╰─➤  cat tmp.js                                                                                                                                                                                                                                           
const os = require('os');

console.log(os.arch());

╭─eli@Johans-MacBook-Pro.local ~/tmp
╰─➤  node tmp.js
x64
```

I would presume you get `arm` there instead of `x64` if you're on M1?

**Comment by mwarger** [comment]
I get x64 back with that.

I tried uname and arch via `exec` as well without luck.  My suspicion is that it's getting that since it's running with node?

**Comment by elitan** [comment]
I'll test on a Raspberry later. Can you test with `process.arch` [link](https://nodejs.org/api/process.html#process_process_arch) if you're quicker than me?

**Comment by mwarger** [comment]
That also returns x64.  I also ran this on node v15 which should be running natively on M1 and also returns x64.

**Comment by mwarger** [comment]
Leaving this here for later: https://stackoverflow.com/questions/65346260/get-real-architecture-of-m1-mac-regardless-of-rosetta

**Comment by mwarger** [comment]
@elitan Please see the PR...  local environment up and working.  If you want any changes let me know.

![image](https://user-images.githubusercontent.com/686823/112534191-c5db2f00-8d78-11eb-89ec-2747e231602a.png)

**Comment by mwarger** [comment]
Additionally, this is likely only an issue until hasura updates their app to support M1 (issue here: https://github.com/hasura/graphql-engine/issues/6337).

**Comment by icybergenome** [comment]
@elitan **nhost/hasura-backend-plus** and **nhost_nhost-api** are not Apple M1 chip supported. It is causing failure in API and initial migration scripts.
![image](https://user-images.githubusercontent.com/12095716/126961221-e2f961f8-e167-4146-b8c9-1857dbb7bcc3.png)

**Comment by elitan** [comment]
> @elitan **nhost/hasura-backend-plus** and **nhost_nhost-api** are not Apple M1 chip supported. It is causing failure in API and initial migration scripts.
> ![image](https://user-images.githubusercontent.com/12095716/126961221-e2f961f8-e167-4146-b8c9-1857dbb7bcc3.png)

Are you sure this is related to M1? What's in the error logs?

**Comment by icybergenome** [comment]
@elitan This is the error when I try to run `nhost dev`
![image](https://user-images.githubusercontent.com/12095716/126963597-b5351a39-56e7-476e-8a6d-27749f80d3c0.png)

**docker: nhost-hbp**:
```
No custom storage security rules found.

(node:48) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432

at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)

(Use `node --trace-warnings ...` to show where the warning was created)

(node:48) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

(node:48) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

No custom storage security rules found.

(node:48) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432

at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)

(Use `node --trace-warnings ...` to show where the warning was created)

(node:48) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

(node:48) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

No custom storage security rules found.

(node:48) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432

at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)

(Use `node --trace-warnings ...` to show where the warning was created)

(node:48) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

(node:48) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

No custom storage security rules found.

(node:48) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432

at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)

(Use `node --trace-warnings ...` to show where the warning was created)

(node:48) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

(node:48) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

No custom storage security rules found.

(node:49) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432

at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)

(Use `node --trace-warnings ...` to show where the warning was created)

(node:49) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

(node:49) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by elitan** [comment]
Do you have a docker-compose file so I could try to reproduce the issue on my end?

**Comment by icybergenome** [comment]
@elitan I used `nhost dev` command to run the container. Here is my config.yaml file:
```
version: 2
metadata_directory: metadata
hasura_graphql_version: v1.3.3
hasura_graphql_port: 8080
hasura_graphql_admin_secret: 123456
hasura_backend_plus_version: v2.5.0
hasura_backend_plus_port: 9001
postgres_version: '12.0'
postgres_port: 5432
postgres_user: postgres
postgres_password: postgres
minio_port: 9000
api_port: 4000
env_file: ../.env.development
provider_success_redirect: 'http://localhost:3000'
provider_failure_redirect: 'http://localhost:3000/login-fail'
google_enable: false
google_client_id: ''
google_client_secret: ''
github_enable: false
github_client_id: ''
github_client_secret: ''
facebook_enable: false
facebook_client_id: ''
facebook_client_secret: ''
linkedin_enable: false
linkedin_client_id: ''
linkedin_client_secret: ''
hasura_graphql_engine: fedormelexin/graphql-engine-arm64
```

**Comment by icybergenome** [comment]
@elitan any idea? As we are still stuck in this issue and we had to use live url and its really effecting our productivity.

**Comment by elitan** [comment]
> @elitan any idea? As we are still stuck in this issue and we had to use live url and it's really effecting our productivity.

Sorry for being a bit slow here.

> No custom storage security rules found.

it looks like it complains about the storage rules.

Can you check your `nhost/custom/storage-rules` folder of your repository? Ideally, there should be a file `rules.yaml` inside that folder.

You can add the following content for `rules.yaml` to get started if it does not exist: https://github.com/nhost/hasura-backend-plus/blob/master/custom/storage-rules/rules.yaml

**Comment by icybergenome** [comment]
@elitan Thanks for responding. I did place `rules.yaml` it removed the warning however rest of the error remains as:
```
(node:49) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

(node:48) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432

at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16)

(Use `node --trace-warnings ...` to show where the warning was created)

(node:48) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)

(node:48) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.


yarn run v1.22.5
```

Major error is `UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:5432` which is postgresql instance is not accessible to nhost network, although its running.
![image](https://user-images.githubusercontent.com/12095716/127296616-5970a46f-74a9-4be8-a2d0-2a3b26d43aa5.png)

**Comment by icybergenome** [comment]
@elitan Btw inside `.nhost` directory following docker-compose file was generated:
```
version: '3.6'
services:
  nhost-postgres:
    container_name: nhost_postgres
    image: 'postgres:12.0'
    ports:
      - '5432:5432'
    restart: always
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
    volumes:
      - './db_data:/var/lib/postgresql/data'
  nhost-graphql-engine:
    container_name: nhost_hasura
    image: 'fedormelexin/graphql-engine-arm64:v1.3.3'
    ports:
      - '8080:8080'
    depends_on:
      - nhost-postgres
    restart: always
    environment:
      HASURA_GRAPHQL_SERVER_PORT: 8080
      HASURA_GRAPHQL_DATABASE_URL: 'postgres://postgres:postgres@nhost-postgres:5432/postgres'
      HASURA_GRAPHQL_ENABLE_CONSOLE: 'false'
      HASURA_GRAPHQL_ENABLED_LOG_TYPES: 'startup, http-log, webhook-log, websocket-log, query-log'
      HASURA_GRAPHQL_ADMIN_SECRET: 123456
      HASURA_GRAPHQL_JWT_SECRET: >-
        {"type":"HS256", "key":
        "d8b437ff350fb875be27b8f6acf2b89d5f3d9ff001168a522bed6a56653673508fd71b94c8d4358e55b395805fead837b6ec9ebdcf01e586d93a09b16d572a89"}
      HASURA_GRAPHQL_MIGRATIONS_SERVER_TIMEOUT: 20
      HASURA_GRAPHQL_NO_OF_RETRIES: 20
      HASURA_GRAPHQL_UNAUTHORIZED_ROLE: public
      NHOST_HASURA_URL: 'http://nhost_hasura:8080/v1/graphql'
      NHOST_WEBHOOK_SECRET: devnhostwebhooksecret
      NHOST_HBP_URL: 'http://nhost_hbp:9001'
      NHOST_CUSTOM_API_URL: 'http://nhost_api:4000'
    env_file:
      - ../.env.development
    command:
      - graphql-engine
      - serve
    volumes:
      - '../nhost/migrations:/hasura-migrations'
  nhost-hasura-backend-plus:
    container_name: nhost_hbp
    image: 'nhost/hasura-backend-plus:v2.5.0'
    ports:
      - '9001:9001'
    depends_on:
      - nhost-graphql-engine
    restart: always
    environment:
      PORT: 9001
      USER_FIELDS: ''
      USER_REGISTRATION_AUTO_ACTIVE: 'true'
      HASURA_GRAPHQL_ENDPOINT: 'http://nhost-graphql-engine:8080/v1/graphql'
      HASURA_ENDPOINT: 'http://nhost-graphql-engine:8080/v1/graphql'
      HASURA_GRAPHQL_ADMIN_SECRET: 123456
      JWT_ALGORITHM: HS256
      JWT_KEY: >-
        d8b437ff350fb875be27b8f6acf2b89d5f3d9ff001168a522bed6a56653673508fd71b94c8d4358e55b395805fead837b6ec9ebdcf01e586d93a09b16d572a89
      AUTH_ACTIVE: 'true'
      AUTH_LOCAL_ACTIVE: 'true'
      REFRESH_TOKEN_EXPIRES: 43200
      JWT_TOKEN_EXPIRES: 15
      S3_ENDPOINT: 'nhost_minio:9000'
      S3_SSL_ENABLED: 'false'
      S3_BUCKET: nhost
      S3_ACCESS_KEY_ID: minioaccesskey123123
      S3_SECRET_ACCESS_KEY: miniosecretkey123123
      LOST_PASSWORD_ENABLE: 'true'
      PROVIDER_SUCCESS_REDIRECT: 'http://localhost:3000'
      PROVIDER_FAILURE_REDIRECT: 'http://localhost:3000/login-fail'
      GOOGLE_ENABLE: 'false'
      GOOGLE_CLIENT_ID: ''
      GOOGLE_CLIENT_SECRET: ''
      GITHUB_ENABLE: 'false'
      GITHUB_CLIENT_ID: ''
      GITHUB_CLIENT_SECRET: ''
      FACEBOOK_ENABLE: 'false'
      FACEBOOK_CLIENT_ID: ''
      FACEBOOK_CLIENT_SECRET: ''
      LINKEDIN_ENABLE: 'false'
      LINKEDIN_CLIENT_ID: ''
      LINKEDIN_CLIENT_SECRET: ''
    env_file:
      - ../.env.development
    volumes:
      - '../nhost/custom:/app/custom'
  minio:
    container_name: nhost_minio
    image: minio/minio
    user: '999:1001'
    restart: always
    volumes:
      - './minio/data:/data'
      - './minio/config:/.minio'
    environment:
      MINIO_ACCESS_KEY: minioaccesskey123123
      MINIO_SECRET_KEY: miniosecretkey123123
    entrypoint: sh
    command: '-c ''mkdir -p /data/nhost && /usr/bin/minio server --address :9000 /data'''
    ports:
      - '9000:9000'
  nhost-api:
    container_name: nhost_api
    build:
      context: ../
      dockerfile: ./.nhost/Dockerfile-api
    environment:
      PORT: 4000
      NHOST_JWT_ALGORITHM: HS256
      NHOST_JWT_KEY: >-
        d8b437ff350fb875be27b8f6acf2b89d5f3d9ff001168a522bed6a56653673508fd71b94c8d4358e55b395805fead837b6ec9ebdcf01e586d93a09b16d572a89
      NHOST_HASURA_URL: 'http://nhost_hasura:8080/v1/graphql'
      NHOST_HASURA_ADMIN_SECRET: 123456
      NHOST_WEBHOOK_SECRET: devnhostwebhooksecret
      NHOST_HBP_URL: 'http://nhost_hbp:9001'
      NHOST_CUSTOM_API_URL: 'http://nhost_api:4000'
    ports:
      - '4000:4000'
    env_file:
      - ../.env.development
    volumes:
      - '../api:/usr/src/app/api'

```

**Comment by icybergenome** [comment]
@elitan Can you please send me a `docker-compose` file which works?

**Comment by icybergenome** [comment]
https://github.com/hasura/graphql-engine/issues/6337
As per hasura issues, the M1 chip issue is still open as of now, However `fedormelexin/graphql-engine-arm64` is still recommended. But in case of `nhost dev` it appends the wrong version v1.3.3 with it. So inside `.nhost/docker-compose.yaml ` I updated the version to `fedormelexin/graphql-engine-arm64`. Also postgres image: `nhost/postgres` should be used.

I made it working with `docker-compose up` and as per docker console all containers are working fine, but for some reason if I visit the URLs like localhost:8080 it display error of 404(i.e. resource not found.)
